maitake/src/wait/dataqueue.rs

```diff
@@ -0,0 +1,984 @@
+    /// real spinlocks.
+    queue: Mutex<List<Waiter<K, V>>>,
+
+    _key: PhantomData<K>,
+    _val: PhantomData<V>,
+}
+
+impl<K: PartialEq, V> Debug for DataWaitQueue<K, V> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("DataWaitQueue")
+            .field("state", &self.state)
+            .field("queue", &self.queue)
+            .field("key", &fmt::display(core::any::type_name::<K>()))
+            .field("val", &fmt::display(core::any::type_name::<V>()))
+            .finish()
+    }
+}
+
+/// Future returned from [`DataWaitQueue::wait()`].
+///
+/// This future is fused, so once it has completed, any future calls to poll
+/// will immediately return [`Poll::Ready`].
+#[derive(Debug)]
+#[pin_project(PinnedDrop)]
+#[must_use = "futures do nothing unless `.await`ed or `poll`ed"]
+pub struct Wait<'a, K: PartialEq, V> {
+    /// The [`DataWaitQueue`] being waited on from.
+    queue: &'a DataWaitQueue<K, V>,
+
+    /// Entry in the wait queue linked list.
+    #[pin]
+    waiter: Waiter<K, V>,
+}
+
+/// A waiter node which may be linked into a wait queue.
+#[repr(C)]
+#[pin_project]
+struct Waiter<K: PartialEq, V> {
+    /// The intrusive linked list node.
+    ///
+    /// This *must* be the first field in the struct in order for the `Linked`
+    /// implementation to be sound.
+    #[pin]
+    node: UnsafeCell<Node<K, V>>,
+
+    /// The future's state.
+    state: WaitStateBits,
+
+    key: K,
+    val: UnsafeCell<Option<V>>,
+}
+
+impl<K: PartialEq, V> Debug for Waiter<K, V> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Waiter")
+            .field("node", &self.node)
+            .field("state", &self.state)
+            .field("key", &fmt::display(core::any::type_name::<K>()))
+            .field("val", &fmt::display(core::any::type_name::<V>()))
+            .finish()
+    }
+}
+
+#[repr(C)]
+struct Node<K: PartialEq, V> {
+    /// Intrusive linked list pointers.
+    ///
+    /// # Safety
+    ///
+    /// This *must* be the first field in the struct in order for the `Linked`
+    /// impl to be sound.
+    links: list::Links<Waiter<K, V>>,
+
+    /// The node's waker
+    waker: Wakeup,
+
+    // This type is !Unpin due to the heuristic from:
+    // <https://github.com/rust-lang/rust/pull/82834>
+    _pin: PhantomPinned,
+}
+
+impl<K: PartialEq, V> Debug for Node<K, V> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Node")
+            .field("links", &self.links)
+            .field("waker", &self.waker)
+            .finish()
+    }
+}
+
+bitfield! {
+    #[derive(Eq, PartialEq)]
+    struct QueueState<usize> {
+        /// The queue's state.
+        const STATE: State;
+    }
+}
+
+bitfield! {
+    #[derive(Eq, PartialEq)]
+    struct WaitStateBits<usize> {
+        /// The waiter's state.
+        const STATE: WaitState;
+    }
+}
```

Comment:
since these are _just_ enums, and don't also contain a count of `wake_all` calls, it seems like we don't need the bitfield and can just convert the `AtomicUsize` directly into a `State` (for the queue), or just store the enum in the waiter directly (for the waiter).

Reply:
Yeah, makes sense! I'll address this in the next chunk of work.